File: naked-generator/Primary/c/Primary.c
#include <string.h>
#include <assert.h>
#include <stdio.h>

#include "Primary.h"

/* Primary_TLM_STATUS */
void serialize_Primary_TLM_STATUS(Primary_Tlm_Status tlm_status, Primary_Race_Type race_type, uint8_t driver, uint8_t circuit, uint8_t* buffer, size_t buf_len) {
    Primary_TLM_STATUS primary_tlm_status = { tlm_status, race_type, driver, circuit };
	assert(buf_len >= sizeof(Primary_TLM_STATUS));
	memcpy(buffer, &primary_tlm_status, sizeof(Primary_TLM_STATUS));
} 
void deserialize_Primary_TLM_STATUS(uint8_t* buffer, size_t buf_len, Primary_TLM_STATUS* primary_tlm_status) {
	assert(buf_len >= sizeof(Primary_TLM_STATUS));
	memcpy(primary_tlm_status, buffer, sizeof(Primary_TLM_STATUS));
}

/* Primary_CAR_STATUS */
void serialize_Primary_CAR_STATUS(Primary_Car_Status car_status, Primary_Inverter_Status inverter_l, Primary_Inverter_Status inverter_r, uint8_t* buffer, size_t buf_len) {
    Primary_CAR_STATUS primary_car_status = { car_status, inverter_l, inverter_r };
	assert(buf_len >= sizeof(Primary_CAR_STATUS));
	memcpy(buffer, &primary_car_status, sizeof(Primary_CAR_STATUS));
} 
void deserialize_Primary_CAR_STATUS(uint8_t* buffer, size_t buf_len, Primary_CAR_STATUS* primary_car_status) {
	assert(buf_len >= sizeof(Primary_CAR_STATUS));
	memcpy(primary_car_status, buffer, sizeof(Primary_CAR_STATUS));
}

/* Primary_SET_TLM_STATUS */
void serialize_Primary_SET_TLM_STATUS(Primary_Tlm_Status tlm_status, Primary_Race_Type race_type, uint8_t driver, uint8_t circuit, uint8_t* buffer, size_t buf_len) {
    Primary_SET_TLM_STATUS primary_set_tlm_status = { tlm_status, race_type, driver, circuit };
	assert(buf_len >= sizeof(Primary_SET_TLM_STATUS));
	memcpy(buffer, &primary_set_tlm_status, sizeof(Primary_SET_TLM_STATUS));
} 
void deserialize_Primary_SET_TLM_STATUS(uint8_t* buffer, size_t buf_len, Primary_SET_TLM_STATUS* primary_set_tlm_status) {
	assert(buf_len >= sizeof(Primary_SET_TLM_STATUS));
	memcpy(primary_set_tlm_status, buffer, sizeof(Primary_SET_TLM_STATUS));
}

/* Primary_HV_VOLTAGE */
void serialize_Primary_HV_VOLTAGE(uint16_t pack_voltage, uint16_t bus_voltage, uint16_t max_cell_voltage, uint16_t min_cell_voltage, uint8_t* buffer, size_t buf_len) {
    Primary_HV_VOLTAGE primary_hv_voltage = { pack_voltage, bus_voltage, max_cell_voltage, min_cell_voltage };
	assert(buf_len >= sizeof(Primary_HV_VOLTAGE));
	memcpy(buffer, &primary_hv_voltage, sizeof(Primary_HV_VOLTAGE));
} 
void deserialize_Primary_HV_VOLTAGE(uint8_t* buffer, size_t buf_len, Primary_HV_VOLTAGE* primary_hv_voltage) {
	assert(buf_len >= sizeof(Primary_HV_VOLTAGE));
	memcpy(primary_hv_voltage, buffer, sizeof(Primary_HV_VOLTAGE));
}

/* Primary_HV_CURRENT */
void serialize_Primary_HV_CURRENT(int8_t power, uint16_t current, uint8_t* buffer, size_t buf_len) {
    Primary_HV_CURRENT primary_hv_current = { power, 0x00, current };
	assert(buf_len >= sizeof(Primary_HV_CURRENT));
	memcpy(buffer, &primary_hv_current, sizeof(Primary_HV_CURRENT));
} 
void deserialize_Primary_HV_CURRENT(uint8_t* buffer, size_t buf_len, Primary_HV_CURRENT* primary_hv_current) {
	assert(buf_len >= sizeof(Primary_HV_CURRENT));
	memcpy(primary_hv_current, buffer, sizeof(Primary_HV_CURRENT));
}

/* Primary_HV_TEMP */
void serialize_Primary_HV_TEMP(uint16_t average_temp, uint16_t max_temp, uint16_t min_temp, uint8_t* buffer, size_t buf_len) {
    Primary_HV_TEMP primary_hv_temp = { average_temp, max_temp, min_temp };
	assert(buf_len >= sizeof(Primary_HV_TEMP));
	memcpy(buffer, &primary_hv_temp, sizeof(Primary_HV_TEMP));
} 
void deserialize_Primary_HV_TEMP(uint8_t* buffer, size_t buf_len, Primary_HV_TEMP* primary_hv_temp) {
	assert(buf_len >= sizeof(Primary_HV_TEMP));
	memcpy(primary_hv_temp, buffer, sizeof(Primary_HV_TEMP));
}

/* Primary_HV_ERROR */
void serialize_Primary_HV_ERROR(uint8_t error_code, uint8_t error_index, uint8_t active, uint8_t* buffer, size_t buf_len) {
    Primary_HV_ERROR primary_hv_error = { error_code, error_index, active };
	assert(buf_len >= sizeof(Primary_HV_ERROR));
	memcpy(buffer, &primary_hv_error, sizeof(Primary_HV_ERROR));
} 
void deserialize_Primary_HV_ERROR(uint8_t* buffer, size_t buf_len, Primary_HV_ERROR* primary_hv_error) {
	assert(buf_len >= sizeof(Primary_HV_ERROR));
	memcpy(primary_hv_error, buffer, sizeof(Primary_HV_ERROR));
}

/* Primary_TS_STATUS */
void serialize_Primary_TS_STATUS(Primary_Ts_Status ts_status, uint8_t* buffer, size_t buf_len) {
    Primary_TS_STATUS primary_ts_status = { ts_status };
	assert(buf_len >= sizeof(Primary_TS_STATUS));
	memcpy(buffer, &primary_ts_status, sizeof(Primary_TS_STATUS));
} 
void deserialize_Primary_TS_STATUS(uint8_t* buffer, size_t buf_len, Primary_TS_STATUS* primary_ts_status) {
	assert(buf_len >= sizeof(Primary_TS_STATUS));
	memcpy(primary_ts_status, buffer, sizeof(Primary_TS_STATUS));
}

/* Primary_STEER_STATUS */
void serialize_Primary_STEER_STATUS(Primary_Traction_Control traction_control, Primary_Map map, bool radio_on, uint8_t* buffer, size_t buf_len) {
    Primary_STEER_STATUS primary_steer_status = { traction_control, map, radio_on };
	assert(buf_len >= sizeof(Primary_STEER_STATUS));
	memcpy(buffer, &primary_steer_status, sizeof(Primary_STEER_STATUS));
} 
void deserialize_Primary_STEER_STATUS(uint8_t* buffer, size_t buf_len, Primary_STEER_STATUS* primary_steer_status) {
	assert(buf_len >= sizeof(Primary_STEER_STATUS));
	memcpy(primary_steer_status, buffer, sizeof(Primary_STEER_STATUS));
}

/* Primary_SET_CAR_STATUS */
void serialize_Primary_SET_CAR_STATUS(Primary_Car_Status_Set car_status_set, uint8_t* buffer, size_t buf_len) {
    Primary_SET_CAR_STATUS primary_set_car_status = { car_status_set };
	assert(buf_len >= sizeof(Primary_SET_CAR_STATUS));
	memcpy(buffer, &primary_set_car_status, sizeof(Primary_SET_CAR_STATUS));
} 
void deserialize_Primary_SET_CAR_STATUS(uint8_t* buffer, size_t buf_len, Primary_SET_CAR_STATUS* primary_set_car_status) {
	assert(buf_len >= sizeof(Primary_SET_CAR_STATUS));
	memcpy(primary_set_car_status, buffer, sizeof(Primary_SET_CAR_STATUS));
}

/* Primary_SET_TS_STATUS */
void serialize_Primary_SET_TS_STATUS(Primary_Ts_Status_Set ts_status_set, uint8_t* buffer, size_t buf_len) {
    Primary_SET_TS_STATUS primary_set_ts_status = { ts_status_set };
	assert(buf_len >= sizeof(Primary_SET_TS_STATUS));
	memcpy(buffer, &primary_set_ts_status, sizeof(Primary_SET_TS_STATUS));
} 
void deserialize_Primary_SET_TS_STATUS(uint8_t* buffer, size_t buf_len, Primary_SET_TS_STATUS* primary_set_ts_status) {
	assert(buf_len >= sizeof(Primary_SET_TS_STATUS));
	memcpy(primary_set_ts_status, buffer, sizeof(Primary_SET_TS_STATUS));
}

File: naked-generator/Primary/py/Primary.py
from enum import Enum
from struct import pack, unpack
from collections import namedtuple


class Secondary_Sync_State(Enum):
    MAX_START = 0
    MAX_END = 1
    MIN_START = 2
    MIN_END = 3


class Secondary_Pedal(Enum):
    ACCELERATOR = 0
    BRAKE = 1
    ALL = 2


class Tlm_Status(Enum):
    ON = 0
    OFF = 1


class Race_Type(Enum):
    ACCELERATION = 0
    SKIDPAD = 1
    AUTOCROSS = 2
    ENDURANCE = 3


class Car_Status(Enum):
    IDLE = 0
    SETUP = 1
    RUN = 2


class Inverter_Status(Enum):
    OFF = 0
    IDLE = 1
    ON = 2


class Ts_Status(Enum):
    OFF = 0
    PRECHARGE = 1
    ON = 2
    FATAL = 3


class Traction_Control(Enum):
    OFF = 0
    SLIP_CONTROL = 1
    TORQUE_VECTORING = 2
    COMPLETE = 3


class Map(Enum):
    R = 0
    D20 = 1
    D40 = 2
    D60 = 3
    D80 = 4
    D100 = 5


class Car_Status_Set(Enum):
    IDLE = 0
    RUN = 1


class Ts_Status_Set(Enum):
    OFF = 0
    ON = 1

# TlmStatus
class TlmStatus:
    struct = namedtuple("TlmStatus_struct", "tlm_status, race_type, driver, circuit", rename=True)
    schema = "<bbbb"
    
    @staticmethod
    def serialize(tlm_status, race_type, driver, circuit) -> bytes:
        return pack(TlmStatus.schema, tlm_status, race_type, driver, circuit)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "TlmStatus.struct":
        return TlmStatus.struct._make(unpack(TlmStatus.schema, buffer))

# CarStatus
class CarStatus:
    struct = namedtuple("CarStatus_struct", "car_status, inverter_l, inverter_r", rename=True)
    schema = "<bbb"
    
    @staticmethod
    def serialize(car_status, inverter_l, inverter_r) -> bytes:
        return pack(CarStatus.schema, car_status, inverter_l, inverter_r)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "CarStatus.struct":
        return CarStatus.struct._make(unpack(CarStatus.schema, buffer))

# SetTlmStatus
class SetTlmStatus:
    struct = namedtuple("SetTlmStatus_struct", "tlm_status, race_type, driver, circuit", rename=True)
    schema = "<bbbb"
    
    @staticmethod
    def serialize(tlm_status, race_type, driver, circuit) -> bytes:
        return pack(SetTlmStatus.schema, tlm_status, race_type, driver, circuit)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "SetTlmStatus.struct":
        return SetTlmStatus.struct._make(unpack(SetTlmStatus.schema, buffer))

# HvVoltage
class HvVoltage:
    struct = namedtuple("HvVoltage_struct", "pack_voltage, bus_voltage, max_cell_voltage, min_cell_voltage", rename=True)
    schema = "<bbbb"
    
    @staticmethod
    def serialize(pack_voltage, bus_voltage, max_cell_voltage, min_cell_voltage) -> bytes:
        return pack(HvVoltage.schema, pack_voltage, bus_voltage, max_cell_voltage, min_cell_voltage)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "HvVoltage.struct":
        return HvVoltage.struct._make(unpack(HvVoltage.schema, buffer))

# HvCurrent
class HvCurrent:
    struct = namedtuple("HvCurrent_struct", "power, current", rename=True)
    schema = "<bbb"
    
    @staticmethod
    def serialize(power, current) -> bytes:
        return pack(HvCurrent.schema, power, 0x00, current)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "HvCurrent.struct":
        return HvCurrent.struct._make(unpack(HvCurrent.schema, buffer))

# HvTemp
class HvTemp:
    struct = namedtuple("HvTemp_struct", "average_temp, max_temp, min_temp", rename=True)
    schema = "<bbb"
    
    @staticmethod
    def serialize(average_temp, max_temp, min_temp) -> bytes:
        return pack(HvTemp.schema, average_temp, max_temp, min_temp)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "HvTemp.struct":
        return HvTemp.struct._make(unpack(HvTemp.schema, buffer))

# HvError
class HvError:
    struct = namedtuple("HvError_struct", "error_code, error_index, active", rename=True)
    schema = "<bbb"
    
    @staticmethod
    def serialize(error_code, error_index, active) -> bytes:
        return pack(HvError.schema, error_code, error_index, active)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "HvError.struct":
        return HvError.struct._make(unpack(HvError.schema, buffer))

# TsStatus
class TsStatus:
    struct = namedtuple("TsStatus_struct", "ts_status", rename=True)
    schema = "<b"
    
    @staticmethod
    def serialize(ts_status) -> bytes:
        return pack(TsStatus.schema, ts_status)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "TsStatus.struct":
        return TsStatus.struct._make(unpack(TsStatus.schema, buffer))

# SteerStatus
class SteerStatus:
    struct = namedtuple("SteerStatus_struct", "traction_control, map, radio_on", rename=True)
    schema = "<bbb"
    
    @staticmethod
    def serialize(traction_control, map, radio_on) -> bytes:
        return pack(SteerStatus.schema, traction_control, map, radio_on)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "SteerStatus.struct":
        return SteerStatus.struct._make(unpack(SteerStatus.schema, buffer))

# SetCarStatus
class SetCarStatus:
    struct = namedtuple("SetCarStatus_struct", "car_status_set", rename=True)
    schema = "<b"
    
    @staticmethod
    def serialize(car_status_set) -> bytes:
        return pack(SetCarStatus.schema, car_status_set)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "SetCarStatus.struct":
        return SetCarStatus.struct._make(unpack(SetCarStatus.schema, buffer))

# SetTsStatus
class SetTsStatus:
    struct = namedtuple("SetTsStatus_struct", "ts_status_set", rename=True)
    schema = "<b"
    
    @staticmethod
    def serialize(ts_status_set) -> bytes:
        return pack(SetTsStatus.schema, ts_status_set)
    
    @staticmethod
    def deserialize(buffer: bytes) -> "SetTsStatus.struct":
        return SetTsStatus.struct._make(unpack(SetTsStatus.schema, buffer))
